### ISSUE TITLE
Bug in pillar_encoder.py

### ISSUE THREAD

**Issue by poddar414** [open-mmlab/mmdetection3d#1504]
Hi,

I am running kitti pointpillars model. And it seems there is bug in features computation. And it is there in below forward function inside the file mmdet3d/models/voxel_encoders/pillar_encoder.py. There are 10 features to be computed for each lidar point and and while calculating the last three features, first 3 features gets overwritten. I guess that is not intended and seems to be bug. As a naïve fix " f_center = features[:, :, :3]" should be replaced with f_center = features[:, :, :3].clone() in else part.

Regards
Deepak Poddar

   @force_fp32(out_fp16=True)
    def forward(self, features, num_points, coors, dump_raw_voxel_feat=False):
        """Forward function.

        Args:
            features (torch.Tensor): Point features or raw points in shape
                (N, M, C).
            num_points (torch.Tensor): Number of points in each pillar.
            coors (torch.Tensor): Coordinates of each voxel.

        Returns:
            torch.Tensor: Features of pillars.
        """
        features_ls = [features]
        # Find distance of x, y, and z from cluster center
        if self._with_cluster_center:
            points_mean = features[:, :, :3].sum(
                dim=1, keepdim=True) / num_points.type_as(features).view(
                    -1, 1, 1)
            f_cluster = features[:, :, :3] - points_mean
            features_ls.append(f_cluster)

        # Find distance of x, y, and z from pillar center
        dtype = features.dtype
        if self._with_voxel_center:
            if not self.legacy:
                f_center = torch.zeros_like(features[:, :, :3])
                f_center[:, :, 0] = features[:, :, 0] - (
                    coors[:, 3].to(dtype).unsqueeze(1) * self.vx +
                    self.x_offset)
                f_center[:, :, 1] = features[:, :, 1] - (
                    coors[:, 2].to(dtype).unsqueeze(1) * self.vy +
                    self.y_offset)
                f_center[:, :, 2] = features[:, :, 2] - (
                    coors[:, 1].to(dtype).unsqueeze(1) * self.vz +
                    self.z_offset)
            else:
                f_center = features[:, :, :3]
                f_center[:, :, 0] = f_center[:, :, 0] - (
                    coors[:, 3].type_as(features).unsqueeze(1) * self.vx +
                    self.x_offset)
                f_center[:, :, 1] = f_center[:, :, 1] - (
                    coors[:, 2].type_as(features).unsqueeze(1) * self.vy +
                    self.y_offset)
                f_center[:, :, 2] = f_center[:, :, 2] - (
                    coors[:, 1].type_as(features).unsqueeze(1) * self.vz +
                    self.z_offset)
            features_ls.append(f_center)

**Comment by poddar414** [comment]
![pointPillarsBug](https://user-images.githubusercontent.com/34803608/169957108-881f4e27-9f73-4454-8bfb-d566c27bd675.png)

Here is the snapshot of the issue. here you can observe that the values of f_center[0][0] is same as features_ls[0][0][0][0:3]. Basically calculation of f_center overwrites the first three features which was original x,y,z.

**Comment by Tai-Wang** [comment]
Please refer to the issue #1150 and #1307 .

**Comment by poddar414** [comment]
Sorry for creating duplicate issue.
